// react-currency-masked-input/react-currency-masked-input.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x2, _x3, _x4) { var _again = true; _function: while (_again) { var object = _x2, property = _x3, receiver = _x4; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x2 = parent; _x3 = property; _x4 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var CurrencyMaskedInput = (function (_Component) {
  _inherits(CurrencyMaskedInput, _Component);

  function CurrencyMaskedInput(props) {
    _classCallCheck(this, CurrencyMaskedInput);

    _get(Object.getPrototypeOf(CurrencyMaskedInput.prototype), 'constructor', this).call(this, props);

    this.state = {
      value: props.value
    };

    this.onChange = this.onChange.bind(this);
  }

  _createClass(CurrencyMaskedInput, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var value = nextProps.value;

      // allows the user to update the value after render
      if (this._isValidUpdateValue(value)) {
        this.setState({ value: value });
      }
    }
  }, {
    key: 'onChange',
    value: function onChange(evt) {
      var _this = this;

      var value = this._maskedInputValue(evt.target.value, evt.target.validity);

      this.setState({ value: value }, function () {
        if (_this.props.onChange) {
          // call original callback, if it exists
          _this.props.onChange(evt, value);
        }
      });
    }
  }, {
    key: '_isValidUpdateValue',
    value: function _isValidUpdateValue(value) {
      // A String of numbers, or a number, will have digits. Null or undefined will not.
      var isANumber = String(value).match(/\d/g);

      return Boolean(isANumber);
    }
  }, {
    key: '_maskedInputValue',
    value: function _maskedInputValue(value) {
      var validity = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

      // a falsy value with "good" input indicates the user is clearing the text,
      // so allow them to.
      if (!value && !validity.badInput) {
        return null;
      }

      // extract digits. if no digits, fill in a zero.
      var digits = value.match(/\d/g) || ['0'];

      // zero-pad a one-digit input
      if (digits.length === 1) {
        digits.unshift('0');
      }

      // add a decimal point
      digits.splice(digits.length - 2, 0, '.');

      // make a number with 2 decimal points
      return Number(digits.join('')).toFixed(2);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2['default'].createElement('input', _extends({
        ref: function () {
          return _this2.value = _this2.state.value;
        },
        type: 'number',
        pattern: '\\d*'
      }, this.props, {
        value: this.state.value,
        onChange: this.onChange
      }));
    }
  }]);

  return CurrencyMaskedInput;
})(_react.Component);

exports['default'] = CurrencyMaskedInput;

CurrencyMaskedInput.propTypes = {
  onChange: _react.PropTypes.func,
  value: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string])
};

CurrencyMaskedInput.defaultProps = {
  value: null
};
module.exports = exports['default'];
